File: src/packages/sdk/src/runtime/context.ts
/**
 * EvalAI Runtime Context - Layer 1 Foundation
 *
 * Execution context management for specifications.
 * Provides clean isolation and proper resource management.
 */

import type { EvalContext, EvalOptions } from "./types";

/**
 * Create a new execution context
 */
export function createContext<TInput = string>(
  input: TInput,
  metadata?: Record<string, unknown>,
  options?: EvalOptions,
): EvalContext & { input: TInput } {
  return {
    input: input as string & TInput,
    metadata,
    options,
  };
}

/**
 * Merge contexts with proper precedence
 * Later contexts override earlier ones
 */
export function mergeContexts(
  base: EvalContext,
  ...overrides: Partial<EvalContext>[]
): EvalContext {
  // Ensure base has a valid input
  if (!base.input) {
    throw new Error("Base context must have a valid input");
  }

  const merged = overrides.reduce(
    (merged, override) => ({
      input: override.input ?? merged.input,
      metadata: {
        ...merged.metadata,
        ...override.metadata,
      },
      options: override.options
        ? {
            ...merged.options,
            ...override.options,
          }
        : merged.options,
    }),
    base,
  );

  // Type assertion since we've ensured input exists
  return merged as EvalContext;
}

/**
 * Clone a context for safe modification
 */
export function cloneContext(context: EvalContext): EvalContext {
  return {
    input: context.input,
    metadata: { ...context.metadata },
    options: context.options ? { ...context.options } : undefined,
  };
}

/**
 * Validate context structure
 */
export function validateContext(context: EvalContext): void {
  if (!context || typeof context !== "object") {
    throw new Error("Context must be an object");
  }

  if (typeof context.input !== "string") {
    throw new Error("Context input must be a string");
  }

  if (context.metadata && typeof context.metadata !== "object") {
    throw new Error("Context metadata must be an object");
  }

  if (context.options && typeof context.options !== "object") {
    throw new Error("Context options must be an object");
  }
}

File: src/packages/sdk-python/src/evalgate_sdk/runtime/context.py
"""EvalGate Runtime Context — Layer 1 Foundation.

Execution context management for specifications.
Port of ``runtime/context.ts``.
"""

from __future__ import annotations

from typing import Any


def create_runtime_context(
    input: str,
    metadata: dict[str, Any] | None = None,
    options: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Create a new execution context."""
    return {
        "input": input,
        "metadata": metadata or {},
        "options": options,
    }


def merge_runtime_contexts(
    base: dict[str, Any],
    *overrides: dict[str, Any],
) -> dict[str, Any]:
    """Merge contexts with proper precedence. Later contexts override earlier ones."""
    if not base.get("input"):
        raise ValueError("Base context must have a valid input")

    merged = dict(base)
    for override in overrides:
        merged = {
            "input": override.get("input") or merged.get("input"),
            "metadata": {**(merged.get("metadata") or {}), **(override.get("metadata") or {})},
            "options": (
                {**(merged.get("options") or {}), **(override.get("options") or {})}
                if override.get("options")
                else merged.get("options")
            ),
        }
    return merged


def clone_runtime_context(context: dict[str, Any]) -> dict[str, Any]:
    """Clone a context for safe modification."""
    return {
        "input": context.get("input"),
        "metadata": dict(context.get("metadata") or {}),
        "options": dict(context.get("options") or {}) if context.get("options") else None,
    }


def validate_runtime_context(context: Any) -> None:
    """Validate context structure."""
    if not isinstance(context, dict):
        raise TypeError("Context must be a dict")

    if not isinstance(context.get("input"), str):
        raise TypeError("Context input must be a string")

    if context.get("metadata") is not None and not isinstance(context.get("metadata"), dict):
        raise TypeError("Context metadata must be a dict")

    if context.get("options") is not None and not isinstance(context.get("options"), dict):
        raise TypeError("Context options must be a dict")
